packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -433,133 +666,236 @@ fn array_to_user_warnings(array: mv8::Value<'_>) -> Vec<UserWarning> {
             if !errors.is_empty() {
                 return Some(Error::User(errors));
             }
         }
     }
 
-    if let Ok(mv8::Value::String(e)) = return_val.get("pkg_error") {
-        // TODO: Don't silently ignore non-string, non-null-or-undefined errors
-        //       (try to convert error value to JSON string and return as error?).
-        return Some(Error::Package(PackageError(e.to_string())));
+    let pkg_error = match new_js_string(scope, "pkg_error") {
+        Ok(pkg_error) => pkg_error,
+        Err(err) => return Some(err),
+    };
+
+    let runner_error = match new_js_string(scope, "runner_error") {
+        Ok(runner_error) => runner_error,
+        Err(err) => return Some(err),
+    };
+
+    if let Some(err) = return_val.get(scope, pkg_error.into()) {
+        // Even though rusty_v8 returns an Option, if the object does not have the property the
+        // result will be `Some(undefined)` rather than `None`
+        if !err.is_undefined() {
+            let err: v8::Local<'s, v8::String> = if let Some(err) = err.to_string(scope) {
+                err
+            } else {
+                return Some(Error::V8("Could not convert err to String".to_string()));
+            };
+
+            // TODO: Don't silently ignore non-string, non-null-or-undefined errors
+            //       (try to convert error value to JSON string and return as error?).
+            return Some(Error::Package(PackageError(
+                err.to_rust_string_lossy(scope),
+            )));
+        }
     }
 
-    if let Ok(mv8::Value::String(e)) = return_val.get("runner_error") {
-        // TODO: Don't ignore non-string, non-null-or-undefined errors
-        return Some(Error::Embedded(e.to_string()));
+    if let Some(err) = return_val.get(scope, runner_error.into()) {
+        // Even though rusty_v8 returns an Option, if the object does not have the property the
+        // result will be `Some(undefined)` rather than `None`
+        if !err.is_undefined() {
+            let err: v8::Local<'s, v8::String> = if let Some(err) = err.to_string(scope) {
+                err
+            } else {
+                return Some(Error::V8("Could not convert err to String".to_string()));
+            };
+
+            // TODO: Don't ignore non-string, non-null-or-undefined errors
+            return Some(Error::Embedded(err.to_rust_string_lossy(scope)));
+        }
     }
 
     None
 }
 
-fn get_user_warnings(_mv8: &MiniV8, return_val: &mv8::Object<'_>) -> Option<Vec<UserWarning>> {
-    if let Ok(warnings) = return_val.get::<&str, mv8::Value<'_>>("user_warnings") {
-        if !(warnings.is_undefined() || warnings.is_null()) {
-            let warnings = array_to_user_warnings(warnings);
+fn get_user_warnings<'s>(
+    scope: &mut v8::HandleScope<'s>,
+    return_val: v8::Local<'s, v8::Object>,
+) -> Result<Option<Vec<UserWarning>>> {
+    let user_warnings = new_js_string(scope, "user_warnings")?;
```

Comment:
Do we really want to error if the user warnings couldn't be generated?

Reply:
It only errors if the string can't be created. rusty_v8 doesn't specify when that can happen but I'd guess the usual culprit: not enough memory.
If we run out of memory we probably have a big issue.

Reply:
It fails, if the string is larger than
- 1Gb on 32-bit
- 2Gb on 64-bit

From the rusty_v8 documentation:
> Allocates a new string from UTF-8 data. Only returns an empty value when length > kMaxLength

see https://stackoverflow.com/a/8974841

Reply:
Ok, so you think we should just check strings given by users? (like behavior files)

Reply:
Removed the Result in https://github.com/hashintel/hash/pull/414/commits/1a4ca4372eea723a8caf056799d2199a6071fc2b